### PR TITLE
Support V2 GRPC for KServe Model Server

### LINKED ISSUE #2107: Custom ModelServer not support inference v2 protocol ?
/kind feature

**Describe the solution you'd like**
When will  custom modelServer supporte v2 protocol  ?  Thanks  .

**Anything else you would like to add:**
image

### PR DESCRIPTION
**Pull request by Suresh-Nakkeran**
**What this PR does / why we need it**:
Add gRPC support for KServe custom model server

**Which issue(s) this PR fixes**:
Fixes #2107 

**Type of changes**
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

**Feature/Issue validation/testing**:
- [x] gRPC custom predictor
- [x] gRPC custom transformer with gRPC predictor
- [x] HTTP custom transformer with gRPC predictor

**Special notes for your reviewer**:

**Release note**:
```
Enable gRPC model server with `--enable_grpc`,  by default it is true.
Configure max gRPC processing threads with `--max_threads`, by default it is 4.
```

### CODE REVIEW

File: python/custom_model/model_grpc.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# Copyright 2021 The KServe Authors.
```
Comment (yuzisun) [inline python/custom_model/model_grpc.py]:
```suggestion
# Copyright 2022 The KServe Authors.
```

File: python/kserve/kserve/grpc/grpc_predict_v2.proto [hunk 2]
```diff
@@ -0,0 +1,362 @@
+// Copyright 2020 kubeflow.org.
```
Comment (yuzisun) [inline python/kserve/kserve/grpc/grpc_predict_v2.proto]:
```suggestion
# Copyright 2022 The KServe Authors.
```

File: docs/samples/kafka/image_transformer/__main__.py [hunk 3]
```diff
@@ -27,4 +28,6 @@
 if __name__ == "__main__":
     transformer = ImageTransformer(args.model_name, predictor_host=args.predictor_host)
     server = kserve.ModelServer()
-    server.start(models=[transformer])
+    asyncio.run(
```
Comment (yuzisun) [inline docs/samples/kafka/image_transformer/__main__.py]: This seems like a breaking change for user, can we do `asyncio` inside `server.start`?

File: test/e2e/custom/test_custom_model_grpc.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def test_custom_model_grpc():
+    service_name = "custom-model"
+
+    predictor = V1beta1PredictorSpec(
+        containers=[
+            V1Container(
+                name="kserve-container",
+                image="suryasmart619/custom-model-grpc:v1", #TODO: Use Kserve repository
```
Comment (yuzisun) [inline test/e2e/custom/test_custom_model_grpc.py]: I think we can build the `CustomModel` as e2e test image like how did with transformer
https://github.com/kserve/kserve/blob/master/test/scripts/gh-actions/build-server-runtimes.sh#L67

File: python/kserve/kserve/model.py [hunk 5]
```diff
@@ -169,9 +170,7 @@ def postprocess(self, response: Union[Dict, ModelInferResponse]) -> Dict:
         :param response: Dict|ModelInferResponse passed from predict handler
         :return: Dict
         """
-        if isinstance(response, ModelInferResponse):
```
Comment (yuzisun) [inline python/kserve/kserve/model.py:231]: need to check whether the original request is REST or gRPC by examining the headers

File: python/kserve/kserve/handlers/dataplane.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
+import inspect
+
+from kserve.model_repository import ModelRepository
+from kserve.utils.utils import generate_uuid
```
Comment (yuzisun) [inline python/kserve/kserve/handlers/dataplane.py]: not used

File: python/kserve/kserve/model_server.py [hunk 7]
```diff
@@ -136,7 +139,7 @@ def start(self, models: Union[List[Model], Dict[str, Deployment]], nest_asyncio:
             import nest_asyncio
             nest_asyncio.apply()
 
-        tornado.ioloop.IOLoop.current().start()
```
Comment (yuzisun) [inline python/kserve/kserve/model_server.py]: first start grpc server and then tornado

File: docs/samples/kafka/image_transformer/__main__.py (hunk 3 above)
Comment (yuzisun) [inline docs/samples/kafka/image_transformer/__main__.py]: So you can't execute `asyncio.run` inside the `server.start`?

File: python/kserve/kserve/grpc/server.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        self,
+        port: int,
+        data_plane: DataPlane
+    ):
+        self._port = port
+        self._data_plane = data_plane
+
+    def _create_server(self):
+        self._inference_servicer = InferenceServicer(self._data_plane)
+        self._server = aio.server(futures.ThreadPoolExecutor(max_workers=5))
```
Comment (yuzisun) [inline python/kserve/kserve/grpc/server.py]: why setting to 5? can we make it configurable?

File: python/kserve/kserve/grpc/server.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        self._create_server()
+
+        await self._server.start()
+        logging.info(
+            "gRPC server running on "
+            f"[::]:{self._port}"
+        )
+        await self._server.wait_for_termination()
+
+    async def stop(self, sig: int = None):
```
Comment (yuzisun) [inline python/kserve/kserve/grpc/server.py:57]: `sig` is not used here

File: test/e2e/custom/test_custom_model_grpc.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+@pytest.mark.fast
+def test_custom_model_grpc():
+    service_name = "custom-model"
+
+    predictor = V1beta1PredictorSpec(
+        containers=[
+            V1Container(
+                name="kserve-container",
+                image="suryasmart619/custom-model-grpc:v1",  # TODO: Use Kserve repository
```
Comment (yuzisun) [inline test/e2e/custom/test_custom_model_grpc.py]: We can build the custom model image in kserve CI

File: python/kserve/kserve/utils/utils.py [hunk 11]
```diff
@@ -116,3 +118,20 @@ def create_response_cloudevent(model_name: str, body: Union[Dict, CloudEvent], r
...
     return event_headers, event_body
+
+
+def generate_uuid() -> str:
+    return str(uuid.uuid4())
+
+
+def to_headers(context: ServicerContext) -> Dict[str, str]:
+    metadata = context.invocation_metadata()
+    if hasattr(context, "trailing_metadata"):
```
Comment (yuzisun) [inline python/kserve/kserve/utils/utils.py:136]: what's `trailing_metadata` ?

File: .github/workflows/python-test.yml [hunk 12]
```diff
@@ -23,7 +23,7 @@ jobs:
         uses: py-actions/flake8@v1
         with:
           max-line-length: "120"
-          exclude: "docs/samples/v1beta1/onnx/assets/*.py,python/kserve/test/test_v1*.py,python/kserve/kserve/__init__.py,python/kserve/test/test_knative*.py"
+          exclude: "docs/samples/v1beta1/onnx/assets/*.py,python/kserve/test/test_v1*.py,python/kserve/kserve/__init__.py,python/kserve/test/test_knative*.py,python/kserve/kserve/grpc/grpc_predict_v2*.py"
```
Comment (yuzisun) [inline .github/workflows/python-test.yml:26]: Why excluding the grpc test?

Comment (Suresh-Nakkeran) [reply .github/workflows/python-test.yml:26]: @yuzisun  Added to skip lint issue with proto files

File: python/kserve/kserve/grpc/server.py (hunk 8 above)
Comment (Suresh-Nakkeran) [reply python/kserve/kserve/grpc/server.py]: I will update with param max_asyncio_workers.

File: python/custom_model/model_grpc.py [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+        result = values.tolist()
+        return {
+            "id": generate_uuid(),
+            "model_name": request.model_name,
+            "outputs": [
+                {
+                    "data": result[0],
+                    "datatype": "FP32",
+                    "name": req.name,
+                    "shape": req.shape
```
Comment (yuzisun) [inline python/custom_model/model_grpc.py]: Here should use the response's data shape

File: python/custom_model/model_grpc.py [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+            # Input follows the Tensorflow V1 HTTP API for binary values
+            # https://www.tensorflow.org/tfx/serving/api_rest#encoding_binary_values
+            data = inputs[0]["image"]["b64"]
+
+            raw_img_data = base64.b64decode(data)
+        elif isinstance(request, ModelInferRequest):
+            req = request.inputs[0]
+            fields = req.contents.ListFields()
+            _, field_value = fields[0]
```
Comment (yuzisun) [inline python/custom_model/model_grpc.py]: we can directly access the `bytes_contents` field here?

File: python/kserve/kserve/model_server.py [hunk 15]
```diff
@@ -155,7 +158,7 @@ def start(self, models: Union[List[Model], Dict[str, Deployment]], nest_asyncio:
...
+        await self._grpc_server.start(self.max_asyncio_workers)
```
Comment (yuzisun) [inline python/kserve/kserve/model_server.py]: use `self.workers`

File: python/kserve/kserve/model_server.py [hunk 16]
```diff
@@ -83,6 +84,8 @@ def __init__(self, http_port: int = args.http_port,
         self.enable_latency_logging = validate_enable_latency_logging(enable_latency_logging)
 
     def create_application(self):
+        dataplane = handlers.DataPlane(model_registry=self.registered_models)
+        self._grpc_server = GRPCServer(port=self.grpc_port, data_plane=dataplane)
```
Comment (yuzisun) [inline python/kserve/kserve/model_server.py]: `create_application` is tornado specific, it is better to separate out grpc server related code.

File: python/kserve/kserve/model_server.py [hunk 17]
```diff
@@ -172,7 +176,7 @@ async def start(self, models: Union[List[Model], Dict[str, Deployment]]) -> None
         )
 
         self._server = uvicorn.Server(cfg)
-        servers = [self._server.serve()]
+        servers = [self._server.serve(), self._grpc_server.start(self.workers)]
```
Comment (ghost) [inline python/kserve/kserve/model_server.py]: @yuzisun @sukumargaonkar and I figured out a way to make the `start()` function *sync again :D 

```
    def start()
      ...

        async def servers_task():
            servers = [self._server.serve(), dummyserver.serve()]
            await asyncio.gather(*servers)

        asyncio.run(servers_task())
```

so we can get rid of all the `asyncio.run()` in the other model servers

File: python/custom_model/runtime.txt [hunk 18]
```diff
@@ -0,0 +1 @@
+python-3.8.14
```
Comment (alexagriffith) [inline python/custom_model/runtime.txt]: this is python 3.8 but the image builds python 3.9, should we keep consistent?

File: python/kserve/kserve/grpc/grpc_predict_v2.proto [hunk 19]
```diff
@@ -0,0 +1,362 @@
+// Copyright 2022 The KServe Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
```
Comment (alexagriffith) [inline python/kserve/kserve/grpc/grpc_predict_v2.proto:4]: just curious, why is this different from https://github.com/kserve/kserve/blob/master/docs/predict-api/v2/grpc_predict_v2.proto ?

**Comment by ghost** [comment]
@yuzisun one thing i just realized is that we didn't put `fastapi` dependency in the `requirements.txt` file. right now the CI could pass because `ray[serve]` uses `fastapi`, I think we should be more explicit and add it to the `requirements.txt`

```
fastapi>=0.85.0
uvicorn>=0.16.0
```

File: python/kserve/kserve/grpc/grpc_predict_v2.proto (hunk 19 above)
Comment (yuzisun) [reply python/kserve/kserve/grpc/grpc_predict_v2.proto:4]: The one in docs directory does not include the model repository extension for load/unload apis

**Comment by yuzisun** [comment]
Awesome work @Suresh-Nakkeran !!

/lgtm
/approve

**Pull request merged**